crates/spirv-std/src/ray_tracing.rs

```diff
@@ -159,3 +159,448 @@ bitflags::bitflags! {
         const SKIP_AABBS = 512;
     }
 }
+
+/// A ray query type which is an opaque object representing a ray traversal.
+#[spirv(ray_query)]
+#[derive(Clone, Default)]
+pub struct RayQuery {
+    _private: (),
+}
+
+impl RayQuery {
+    /// Initialize a ray query object, defining parameters of traversal. After this
+    /// call, a new ray trace can be performed with `ray_query_proceed_khr`. Any
+    /// previous traversal state stored in the object is lost.
+    ///
+    /// - `ray_query` is a pointer to the ray query to initialize.
+    /// - `acceleration_structure` is the descriptor for the acceleration structure
+    ///   to trace into.
+    /// - `ray_flags` contains one or more of the Ray Flag values.
+    /// - `cull_mask` is the mask to test against the instance mask.  Only the 8
+    ///   least-significant bits of `cull_mask` are used by this instruction - other
+    ///   bits are ignored.
+    /// - `ray_origin`, `ray_tmin`, `ray_direction`, and `ray_tmax` control the
+    ///   basic parameters of the ray to be traced.
+    #[spirv_std_macros::gpu_only]
+    #[doc(alias = "OpRayQueryInitializeKHR")]
+    #[inline]
+    #[allow(clippy::too_many_arguments)]
+    pub fn initialize(
+        acceleration_structure: &AccelerationStructure,
+        ray_flags: RayFlags,
+        cull_mask: u32,
+        ray_origin: impl Vector<f32, 3>,
+        ray_tmin: f32,
+        ray_direction: impl Vector<f32, 3>,
+        ray_tmax: f32,
+    ) -> Self {
+        unsafe {
+            let this = Self { _private: () };
```

Comment:
IMO this should be mutable (and a mutable reference passed in `asm!`), but also `MaybeUninit::uninit()` if possible.

I thought opaque types wrapped an `u32` or something, why does this have `()`?

Reply:
Well because using `u32` causes the type to not be able to be constructed in rust because of pointer casts.

```
error: Cannot cast between pointer types
   --> /Users/erin.power/src/rust-gpu2/crates/spirv-std/src/ray_tracing.rs:198:28
    |
198 |             let mut this = Self { _private: 0 };
    |                            ^^^^^^^^^^^^^^^^^^^^
    |
    = note: from: *RayQueryKhr
    = note: to: *u32

error: aborting due to previous error
```

I tried using the same hack that used for `AccelerationStructure`, however this causes some issues as there seem to issues around using `OpTypePointer` with `RayQueryKHR` missing from the type cache. 

```rust
pub fn initialize(
    acceleration_structure: &AccelerationStructure,
    ray_flags: RayFlags,
    cull_mask: u32,
    ray_origin: impl Vector<f32, 3>,
    ray_tmin: f32,
    ray_direction: impl Vector<f32, 3>,
    ray_tmax: f32,
) -> Self {
    unsafe {
        asm! {
            "%ray = OpTypeRayQueryKHR",
            "%ptr = OpTypePointer Generic %ray",
            "%handle = OpVariable %ptr Generic",
            "OpRayQueryInitializeKHR \
            %handle \
            {acceleration_structure} \
            {ray_flags} \
            {cull_mask} \
            {ray_origin} \
            {ray_tmin} \
            {ray_direction} \
            {ray_tmax}",
            "OpReturnValue %handle",
            "%blah = OpLabel",
            acceleration_structure = in(reg) acceleration_structure,
            ray_flags = in(reg) ray_flags.bits(),
            cull_mask = in(reg) cull_mask,
            ray_origin = in(reg) &ray_origin,
            ray_tmin = in(reg) ray_tmin,
            ray_direction = in(reg) &ray_direction,
            ray_tmax = in(reg) ray_tmax,
        }

        loop {}
    }
}
```

Reply:
Did you add the special handling necessary in `spirv_asm` for `OpTypeRayQueryKHR`? The other type instructions get recognized and converted into `SpirvType` instead of simply being emitted.

Either way, `_private: 0` is not what I had in mind, but rather using `MaybeUninit` (if possible).

Reply:
> Either way, _private: 0 is not what I had in mind, but rather using MaybeUninit (if possible).

MaybeUninit doesn't work, I've tried using it before. If you do you get the following error.

```
thread 'rustc' panicked at 'assertion failed: `(left != right)`
  left: `0`,
 right: `0`: TyAndLayout {
    ty: core::mem::MaybeUninit<spirv_std::ray_tracing::RayQuery>,
    layout: Layout {
        fields: Union(
            2,
        ),
        variants: Single {
            index: 0,
        },
        abi: Aggregate {
            sized: true,
        },
        largest_niche: None,
        align: AbiAndPrefAlign {
            abi: Align {
                pow2: 0,
            },
            pref: Align {
                pow2: 3,
            },
        },
        size: Size {
            raw: 0,
        },
    },
}', crates/rustc_codegen_spirv/src/abi.rs:565:13
```